In [3]:
import numpy as np
import ast
import pandas as pd

def get_chain_coverage(pdb_id, chains):
    # Turn chains into a list of strings
    chains = eval(chains)
    coverage = 0 
    for chain in chains:
        atom_file_path = f"pdb_files/{pdb_id}.apdb"
        chain_id = chain
        residue_numbers = set()  # Initialize a set to store unique residue numbers

        try: 
            with open(atom_file_path, 'r') as atom_file:
                for line in atom_file:
                    if line.startswith('ATOM'):
                        current_chain_id = line[21]
                        if current_chain_id == chain_id:
                            residue_number = int(line[22:26].strip())
                            residue_numbers.add(residue_number)  # Add residue number to the set

        except: 
            return 0

        if not residue_numbers:  # Check if the set is empty
            return 0

        coverage += len(residue_numbers) # Increment coverage by the number of unique residue numbers in the set
    #print(f"PDB_ID : {residue_numbers}")
    return coverage

In [4]:
path = "Uniport_human_AF2_Ranked_Chains_Filtered_missing_confidence.xlsx"
df_full = pd.read_excel(path)
df_full

Entry  Reviewed   Entry Name  \
0     Q9NQ94  reviewed   A1CF_HUMAN   
1     Q8N5Z0  reviewed  AADAT_HUMAN   
2     Q2M2I8  reviewed   AAK1_HUMAN   
3     Q9H7C9  reviewed  AAMDC_HUMAN   
4     Q16613  reviewed   SNAT_HUMAN   
...      ...       ...          ...   
7600  P12956  reviewed  XRCC6_HUMAN   
7601  P63104  reviewed  1433Z_HUMAN   
7602  Q6NZY4  reviewed  ZCHC8_HUMAN   
7603  P0DOX5  reviewed   IGG1_HUMAN   
7604  P0DOX7  reviewed    IGK_HUMAN   

                                          Protein names  \
0     APOBEC1 complementation factor (APOBEC1-stimul...   
1     Kynurenine/alpha-aminoadipate aminotransferase...   
2     AP2-associated protein kinase 1 (EC 2.7.11.1) ...   
3     Mth938 domain-containing protein (Adipogenesis...   
4     Serotonin N-acetyltransferase (Serotonin acety...   
...                                                 ...   
7600  X-ray repair cross-complementing protein 6 (EC...   
7601  14-3-3 protein zeta/delta (Protein kinase C in...   
7602  Zinc finger CCHC domain-containing protein 8 (...   
7603  Immunoglobulin gamma-1 heavy chain (Immunoglob...   
7604  Immunoglobulin kappa light chain (Immunoglobul...   

                 Gene Names              Organism  Length  \
0              A1CF ACF ASP  Homo sapiens (Human)     594   
1          AADAT KAT2 KYAT2  Homo sapiens (Human)     425   
2             AAK1 KIAA1048  Homo sapiens (Human)     961   
3     AAMDC C11orf67 PTD015  Homo sapiens (Human)     122   
4                AANAT SNAT  Homo sapiens (Human)     207   
...                     ...                   ...     ...   
7600            XRCC6 G22P1  Homo sapiens (Human)     609   
7601                  YWHAZ  Homo sapiens (Human)     245   
7602                 ZCCHC8  Homo sapiens (Human)     707   
7603                    NaN  Homo sapiens (Human)     449   
7604                    NaN  Homo sapiens (Human)     214   

                                      PDB_Files_Ordered BestRanked  \
0                                              ['2CPD']       2CPD   
1     ['6D0A', '4GEB', '4GE9', '4GE7', '6T8Q', '5EFS...       6D0A   
2                              ['5TE0', '4WSQ', '5L4Q']       5TE0   
3                                      ['2Q4Q', '2AB1']       2Q4Q   
4                                              ['6T80']       6T80   
...                                                 ...        ...   
7600  ['1JEQ', '7SGL', '8AG4', '8AG5', '7AXZ', '6ERG...       1JEQ   
7601  ['4ZDR', '8AH2', '6ZFD', '6ZFG', '7Q16', '3RDH...       4ZDR   
7602  ['7S7C', '7S7B', '7Z4Y', '7Z4Z', '6C90', '5LXR...       7S7C   
7603  ['8DV1', '8DV2', '7CZV', '7CZU', '5O4E', '6BFT...       8DV1   
7604  ['6N35', '6UOE', '6MUB', '6UTE', '6N2X', '5VIG...       6N35   

                Chain Chain_id  \
0           A=223-308    ['A']   
1             A=1-425    ['A']   
2            A=27-365    ['A']   
3           A/B=2-122    ['A']   
4     E/F/G/H=197-207    ['E']   
...               ...      ...   
7600          A=1-609    ['A']   
7601        A/B=1-230    ['A']   
7602        B/F=1-707    ['B']   
7603        D=218-449    ['D']   
7604        K/L=1-213    ['K']   

                                                Missing  \
0     [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
1                                                  [[]]   
2     [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
3                                            [[14, 71]]   
4     [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
...                                                 ...   
7600  [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
7601           [[71, 72, 231, 232, 233, 234, 235, 236]]   
7602  [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
7603  [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
7604                                               [[]]   

                                      Score_missing_res  AvgScore_missing  \
0     {1: 38.84, 2: 34.59, 3: 40.94, 4: 36.54, 5: 40...     

In [56]:
df_full["Coverage"] = df_full.apply(lambda row: get_chain_coverage(row["BestRanked"], row["Chain_id"]), axis=1)

In [60]:
df_full["Percent Covered"] = df_full.apply(lambda row: row["Coverage"] / row["Length"] * 100, axis=1)

### There are 472 out of 7603 PDB files where coverage is > 100 % this could be due to 
 - negative residues present in the pdb file
 - discreptencies in PDB file chain naming (authors vs uniprot) 
 - Post-translational modifcation can introduce additional atoms or residues in the protein structure 
 - different isoforms with varying lenghts may exist
 

In [61]:
rows_greater_than_100 = df_full[df_full["Percent Covered"] > 100]
rows_greater_than_100

Entry  Reviewed   Entry Name  \
1     Q8N5Z0  reviewed  AADAT_HUMAN   
31    Q9P1F3  reviewed  ABRAL_HUMAN   
53    Q9NUN7  reviewed  ACER3_HUMAN   
54    P22303  reviewed   ACES_HUMAN   
67    P24666  reviewed   PPAC_HUMAN   
...      ...       ...          ...   
7576  P68036  reviewed  UB2L3_HUMAN   
7578  Q15819  reviewed  UB2V2_HUMAN   
7579  P07911  reviewed   UROM_HUMAN   
7589  Q9UBQ0  reviewed  VPS29_HUMAN   
7603  P0DOX5  reviewed   IGG1_HUMAN   

                                          Protein names  \
1     Kynurenine/alpha-aminoadipate aminotransferase...   
31    Costars family protein ABRACL (ABRA C-terminal...   
53    Alkaline ceramidase 3 (AlkCDase 3) (Alkaline C...   
54             Acetylcholinesterase (AChE) (EC 3.1.1.7)   
67    Low molecular weight phosphotyrosine protein p...   
...                                                 ...   
7576  Ubiquitin-conjugating enzyme E2 L3 (EC 2.3.2.2...   
7578  Ubiquitin-conjugating enzyme E2 variant 2 (DDV...   
7579  Uromodulin (Tamm-Horsfall urinary glycoprotein...   
7589  Vacuolar protein sorting-associated protein 29...   
7603  Immunoglobulin gamma-1 heavy chain (Immunoglob...   

                           Gene Names              Organism  Length  \
1                    AADAT KAT2 KYAT2  Homo sapiens (Human)     425   
31    ABRACL C6orf115 HSPC280 PRO2013  Homo sapiens (Human)      81   
53                    ACER3 APHC PHCA  Homo sapiens (Human)     267   
54                               ACHE  Homo sapiens (Human)     614   
67                               ACP1  Homo sapiens (Human)     158   
...                               ...                   ...     ...   
7576               UBE2L3 UBCE7 UBCH7  Homo sapiens (Human)     154   
7578                 UBE2V2 MMS2 UEV2  Homo sapiens (Human)     145   
7579                             UMOD  Homo sapiens (Human)     640   
7589            VPS29 DC15 DC7 MDS007  Homo sapiens (Human)     182   
7603                              NaN  Homo sapiens (Human)     449   

                                      PDB_Files_Ordered BestRanked  \
1     ['6D0A', '4GEB', '4GE9', '4GE7', '6T8Q', '5EFS...       6D0A   
31                                             ['2L2O']       2L2O   
53                                     ['6YXH', '6G7O']       6YXH   
54    ['5HQ3', '4BDT', '6O5V', '8DT7', '6U37', '6O4X...       5HQ3   
67    ['4Z99', '4Z9B', '1XWW', '5PNT', '7KH8', '5JNT...       4Z99   
...                                                 ...        ...   
7576  ['4Q5E', '4Q5H', '3SY2', '7V8F', '6CP2', '7OIK...       4Q5E   
7578  ['7BBD', '7BBF', '4ONL', '4ONM', '4ONN', '4NR3...       7BBD   
7579  ['4WRN', '6TQK', '6TQL', '6ZYA', '6ZS5', '7PFP...       4WRN   
7589  ['5GTU', '5WYH', '6XS7', '6XS9', '6XS5', '6XSA...       5GTU   
7603  ['8DV1', '8DV2', '7CZV', '7CZU', '5O4E', '6BFT...       8DV1   

                   Chain    Chain_id  \
1                A=1-425       ['A']   
31                A=1-81       ['A']   
53               A=2-244       ['A']   
54    A=32-578, B=32-579  ['A', 'B']   
67               A=1-158       ['A']   
...                  ...         ...   
7576             C=1-154       ['C']   
7578             A=1-145       ['A']   
7579         A/B=295-610       ['A']   
7589             A=2-182       ['A']   
7603           D=218-449       ['D']   

                                                Missing  \
1                                                  [[]]   
31                                                 [[]]   
53                                                 [[]]   
54    [[1, 2, 3, 4, 260, 261, 262, 263, 264, 494, 49...   
67                                                 [[]]   
...                                                 ...   
7576                                               [[]]   
7578                                               [[]]   
7579  [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1...   
7589                                               [[]]   
7603 

In [63]:
df_full.to_excel("Dataset.xlsx", index = False)